src/scalar_low_impl.h

```diff
@@ -38,8 +38,10 @@ static int secp256k1_scalar_add(secp256k1_scalar *r, const secp256k1_scalar *a,
 
 static void secp256k1_scalar_cadd_bit(secp256k1_scalar *r, unsigned int bit, int flag) {
     if (flag && bit < 32)
-        *r += (1 << bit);
+        *r += ((uint32_t)1 << bit);
 #ifdef VERIFY
+    VERIFY_CHECK(bit < 32);
+    VERIFY_CHECK(((uint32_t)1 << bit) - 1 <= UINT32_MAX - EXHAUSTIVE_TEST_ORDER);
```

Comment:
I don't understand this line. Shouldn't it be `VERIFY_CHECK(((uint32_t)1 << bit) - 1 <= EXHAUSTIVE_TEST_ORDER);`?

Reply:
Or even better, it should check if overflow in `r` has happened (not just in `(uint32_t)1 << bit`.

Reply:
My intention here was to check to see there exists any (valid) value of `r` such that adding `1 << bit` would exceed UINT32_MAX, rather than just checking that this value of `r` doesn't overflow.  So the code as written verifies `(uint32_t)1 + (EXHAUSTIVE_TEST_ORDER -1)` <= UINT32_MAX.

Maybe I shouldn't be the one editing this code. :)

Reply:
My point was that the documentation of this function say that `r` is not allowed to overflow (and that means "overflow as a scalar", not "overflow the underlying integer type"). So I think this is what we should check, and this is consistent with the other implementations. BUT yes, this is already checked, it's just one line below, and I haven't seen it, nevermind!

For your line, we were just not clever enough to figure out what it is supposed to to. In fact Greg has suggested this, now that I read it again...

So I'm fine with this as it is currently but a comment may be useful
```suggestion
    /* Verify that adding (1 << bit) will not overflow any in-range scalar *r by overflowing the underlying uint32_t. /*
    VERIFY_CHECK(((uint32_t)1 << bit) - 1 <= UINT32_MAX - EXHAUSTIVE_TEST_ORDER);
```

Reply:
Done.